jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,369 @@
+ * requests that do not require access to the limited resource.</p>
+ * <p>Requests are resumed in priority order, so that when the
+ * server is under load, and there are many requests suspended to
+ * be processed, high priority request are processed first.
+ * For example, load balancer "ping" requests may have the highest
+ * priority, followed by requests performed by admin users, etc.
+ * so that regardless of the load, "ping" and "admin" requests will
+ * always be able to access the web application.</p>
+ */
+@ManagedObject
+public class QoSHandler extends Handler.Wrapper
+{
+    private static final Logger LOG = LoggerFactory.getLogger(QoSHandler.class);
+
+    private CyclicTimeouts<Entry> timeouts;
+    private List<Queue<Entry>> queues;
+    private int maxRequests;
+    private Duration maxSuspend = Duration.ZERO;
+    private int maxPriority;
+    private Semaphore semaphore;
+
+    public QoSHandler()
+    {
+        super(false);
+    }
+
+    /**
+     * @return the max number of concurrent requests
+     */
+    @ManagedAttribute("The maximum number of concurrent requests")
+    public int getMaxRequests()
+    {
+        return maxRequests;
+    }
+
+    /**
+     * <p>Sets the max number of concurrent requests.</p>
+     * <p>A negative or zero value indicates to calculate
+     * a value based on heuristics, drawn from the number
+     * the size of the server thread pool and/or the number
+     * of CPU cores.</p>
+     *
+     * @param maxRequests the max number of concurrent requests
+     */
+    public void setMaxRequests(int maxRequests)
+    {
+        this.maxRequests = maxRequests;
+    }
+
+    /**
+     * @return the max duration of time a request may stay suspended
+     */
+    public Duration getMaxSuspend()
+    {
+        return maxSuspend;
+    }
+
+    /**
+     * <p>Sets the max duration of time a request may stay suspended.</p>
+     * <p>{@link Duration#ZERO} means that the request may stay suspended forever.</p>
+     *
+     * @param maxSuspend the max duration of time a request may stay suspended
+     */
+    public void setMaxSuspend(Duration maxSuspend)
+    {
+        if (maxSuspend.isNegative())
+            throw new IllegalArgumentException("invalid maxSuspend duration");
+        this.maxSuspend = maxSuspend;
+    }
+
+    /**
+     * @return the max priority of suspended requests
+     */
+    @ManagedAttribute("The maximum priority of suspended requests")
+    public int getMaxPriority()
+    {
+        return maxPriority;
+    }
+
+    /**
+     * <p>Sets the max priority that suspended requests may have.</p>
+     * <p>The value must be {@code 0} or greater, and typically only
+     * few priorities are necessary, say from {@code 0} to {@code 4}
+     * for 5 different priorities.</p>
+     *
+     * @param maxPriority the max priority of suspended requests
+     * @see #getPriority(Request)
+     */
+    public void setMaxPriority(int maxPriority)
```

Comment:
This feels like exposing an implementation detail that we use a fixed size array list.
We could use a sorted data structure that would not need this method.   Hmm but might be worth the efficiency of a final array list of queues? Maybe better as a constructor argument?